File: apps/android/core-financiero/src/main/java/dev/tohure/android_rust_test/adapter/UniffiCoreFinanciero.kt
package dev.tohure.android_rust_test.adapter

import uniffi.core_financiero.Account
import uniffi.core_financiero.TransferRequest
import uniffi.core_financiero.TransferResult
import uniffi.core_financiero.ValidCard
import uniffi.core_financiero.ValidCci
import uniffi.core_financiero.add as coreAdd
import uniffi.core_financiero.calculateItf as coreCalculateItf
import uniffi.core_financiero.coreVersion as coreCoreVersion
import uniffi.core_financiero.decrypt as coreDecrypt
import uniffi.core_financiero.encrypt as coreEncrypt
import uniffi.core_financiero.executeTransfer as coreExecuteTransfer
import uniffi.core_financiero.subtract as coreSubtract
import uniffi.core_financiero.validateCard as coreValidateCard
import uniffi.core_financiero.validateCci as coreValidateCci

/**
 * Habla con `libcore_financiero.so` a través de los bindings de uniffi, que corren sobre
 * JNA. Las llamadas son síncronas y de microsegundos: **no van en corrutinas ni en
 * `Dispatchers.IO`**, salvo en la pantalla de benchmark.
 */
class UniffiCoreFinanciero : CoreFinanciero {
    override fun add(a: String, b: String) = runCatching { coreAdd(a, b) }
    override fun subtract(a: String, b: String) = runCatching { coreSubtract(a, b) }
    override fun calculateItf(amount: String) = runCatching { coreCalculateItf(amount) }
    override fun validateCci(cci: String): Result<ValidCci> = runCatching { coreValidateCci(cci) }
    override fun validateCard(number: String): Result<ValidCard> = runCatching { coreValidateCard(number) }

    override fun encrypt(text: String, keyHex: String, nonceHex: String) =
        runCatching { coreEncrypt(text, keyHex, nonceHex) }

    override fun decrypt(ciphertextHex: String, keyHex: String, nonceHex: String) =
        runCatching { coreDecrypt(ciphertextHex, keyHex, nonceHex) }

    override fun transfer(accounts: List<Account>, request: TransferRequest): Result<TransferResult> =
        runCatching { coreExecuteTransfer(accounts, request) }

    override fun coreVersion(): String = coreCoreVersion()
}

File: apps/ios/CoreFinancieroKit/Adapter/UniffiCoreFinanciero.swift
/// Llama a las funciones globales que emite uniffi.
///
/// **No traduce errores**: propaga el `DomainError` tal cual. Quien lo convierte a texto de
/// usuario es el ViewModel, con `ContractMessages`.
///
/// Los nombres del core se reexportan, no se renombran: una segunda nomenclatura en Swift
/// es una capa que hay que mantener sincronizada a mano y que se desincroniza en la primera
/// regeneración de bindings. El prefijo `CoreFinancieroKit.` en cada llamada desambigua la
/// función global del método de este `struct`, que se llaman igual.
public struct UniffiCoreFinanciero: CoreFinanciero {
    public init() {}

    public func add(a: String, b: String) throws -> String {
        try CoreFinancieroKit.add(a: a, b: b)
    }

    public func subtract(a: String, b: String) throws -> String {
        try CoreFinancieroKit.subtract(a: a, b: b)
    }

    public func calculateItf(amount: String) throws -> String {
        try CoreFinancieroKit.calculateItf(amount: amount)
    }

    public func validateCci(cci: String) throws -> ValidCci {
        try CoreFinancieroKit.validateCci(cci: cci)
    }

    public func validateCard(number: String) throws -> ValidCard {
        try CoreFinancieroKit.validateCard(number: number)
    }

    public func encrypt(text: String, keyHex: String, nonceHex: String) throws -> String {
        try CoreFinancieroKit.encrypt(text: text, keyHex: keyHex, nonceHex: nonceHex)
    }

    public func decrypt(ciphertextHex: String, keyHex: String, nonceHex: String) throws -> String {
        try CoreFinancieroKit.decrypt(ciphertextHex: ciphertextHex, keyHex: keyHex, nonceHex: nonceHex)
    }

    public func transfer(accounts: [Account], request: TransferRequest) throws -> TransferResult {
        try CoreFinancieroKit.executeTransfer(accounts: accounts, request: request)
    }

    public func coreVersion() -> String {
        CoreFinancieroKit.coreVersion()
    }
}
